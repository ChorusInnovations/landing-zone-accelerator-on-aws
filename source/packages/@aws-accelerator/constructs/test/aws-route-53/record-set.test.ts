/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';

import { SynthUtils } from '@aws-cdk/assert';

import { HostedZone, RecordSet, SecurityGroup, VpcEndpoint } from '../../index';

const testNamePrefix = 'Construct(RecordSet): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();
const hostedZoneName = HostedZone.getHostedZoneNameForService('s3-global.accesspoint', stack.region);

const hostedZone = new HostedZone(stack, `TestHostedZone`, {
  hostedZoneName,
  vpcId: 'Test',
});

const securityGroup = new SecurityGroup(stack, 'TestSecurityGroup`', {
  securityGroupName: 'TestSecurityGroup',
  description: `AWS Private Endpoint Zone`,
  vpcId: 'Test',
  tags: [],
});

// Create the interface endpoint
const endpoint = new VpcEndpoint(stack, `TestVpcEndpoint`, {
  vpcId: 'Test',
  vpcEndpointType: cdk.aws_ec2.VpcEndpointType.INTERFACE,
  service: 'ec2',
  subnets: ['Test1', 'Test2'],
  securityGroups: [securityGroup],
  privateDnsEnabled: false,
  policyDocument: new cdk.aws_iam.PolicyDocument({
    statements: [
      new cdk.aws_iam.PolicyStatement({
        sid: 'AccessToTrustedPrincipalsAndResources',
        actions: ['*'],
        effect: cdk.aws_iam.Effect.ALLOW,
        resources: ['*'],
        principals: [new cdk.aws_iam.AnyPrincipal()],
        conditions: {
          StringEquals: {
            'aws:PrincipalOrgID': ['organizationId'],
          },
        },
      }),
    ],
  }),
});

new RecordSet(stack, `TestRecordSet`, {
  type: 'A',
  name: hostedZoneName,
  hostedZone: hostedZone,
  dnsName: endpoint.dnsName,
  hostedZoneId: endpoint.hostedZoneId,
});

/**
 * RecordSet construct test
 */
describe('RecordSet', () => {
  /**
   * Snapshot test
   */
  test(`${testNamePrefix} Snapshot Test`, () => {
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });

  /**
   * Number of HostedZone test
   */
  test(`${testNamePrefix} Hosted zone resource count test`, () => {
    cdk.assertions.Template.fromStack(stack).resourceCountIs('AWS::Route53::HostedZone', 1);
  });

  /**
   * Number of RecordSet test
   */
  test(`${testNamePrefix} RecordSet resource count test`, () => {
    cdk.assertions.Template.fromStack(stack).resourceCountIs('AWS::Route53::RecordSet', 1);
  });

  /**
   * Number of SecurityGroup test
   */
  test(`${testNamePrefix} SecurityGroup resource count test`, () => {
    cdk.assertions.Template.fromStack(stack).resourceCountIs('AWS::EC2::SecurityGroup', 1);
  });

  /**
   * Number of VPCEndpoint test
   */
  test(`${testNamePrefix} VPCEndpoint resource count test`, () => {
    cdk.assertions.Template.fromStack(stack).resourceCountIs('AWS::EC2::VPCEndpoint', 1);
  });

  /**
   * HostedZone resource configuration test
   */
  test(`${testNamePrefix} HostedZone resource configuration test`, () => {
    cdk.assertions.Template.fromStack(stack).templateMatches({
      Resources: {
        TestHostedZone68F306E4: {
          Type: 'AWS::Route53::HostedZone',
          Properties: {
            Name: 's3-global.accesspoint.aws.com',
            VPCs: [
              {
                VPCId: 'Test',
                VPCRegion: {
                  Ref: 'AWS::Region',
                },
              },
            ],
          },
        },
      },
    });
  });

  /**
   * RecordSet resource configuration test
   */
  test(`${testNamePrefix} RecordSet resource configuration test`, () => {
    cdk.assertions.Template.fromStack(stack).templateMatches({
      Resources: {
        TestRecordSetED81F5C1: {
          Type: 'AWS::Route53::RecordSet',
          Properties: {
            AliasTarget: {
              DNSName: {
                'Fn::Select': [
                  1,
                  {
                    'Fn::Split': [
                      ':',
                      {
                        'Fn::Select': [
                          0,
                          {
                            'Fn::GetAtt': ['TestVpcEndpointF7CADE71', 'DnsEntries'],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
              HostedZoneId: {
                'Fn::Select': [
                  0,
                  {
                    'Fn::Split': [
                      ':',
                      {
                        'Fn::Select': [
                          0,
                          {
                            'Fn::GetAtt': ['TestVpcEndpointF7CADE71', 'DnsEntries'],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            },
            HostedZoneId: {
              Ref: 'TestHostedZone68F306E4',
            },
            Name: 's3-global.accesspoint.aws.com',
            Type: 'A',
          },
        },
      },
    });
  });

  /**
   * SecurityGroup resource configuration test
   */
  test(`${testNamePrefix} SecurityGroup resource configuration test`, () => {
    cdk.assertions.Template.fromStack(stack).templateMatches({
      Resources: {
        TestSecurityGroupDA4B5F83: {
          Type: 'AWS::EC2::SecurityGroup',
          Properties: {
            GroupDescription: 'AWS Private Endpoint Zone',
            GroupName: 'TestSecurityGroup',
            Tags: [
              {
                Key: 'Name',
                Value: 'TestSecurityGroup',
              },
            ],
            VpcId: 'Test',
          },
        },
      },
    });
  });

  /**
   * VPCEndpoint resource configuration test
   */
  test(`${testNamePrefix} VPCEndpoint resource configuration test`, () => {
    cdk.assertions.Template.fromStack(stack).templateMatches({
      Resources: {
        TestVpcEndpointF7CADE71: {
          Type: 'AWS::EC2::VPCEndpoint',
          Properties: {
            PolicyDocument: {
              Statement: [
                {
                  Action: '*',
                  Condition: {
                    StringEquals: {
                      'aws:PrincipalOrgID': ['organizationId'],
                    },
                  },
                  Effect: 'Allow',
                  Principal: {
                    AWS: '*',
                  },
                  Resource: '*',
                  Sid: 'AccessToTrustedPrincipalsAndResources',
                },
              ],
              Version: '2012-10-17',
            },
            PrivateDnsEnabled: false,
            SecurityGroupIds: [
              {
                Ref: 'TestSecurityGroupDA4B5F83',
              },
            ],
            ServiceName: {
              'Fn::Join': [
                '',
                [
                  'com.amazonaws.',
                  {
                    Ref: 'AWS::Region',
                  },
                  '.ec2',
                ],
              ],
            },
            SubnetIds: ['Test1', 'Test2'],
            VpcEndpointType: 'Interface',
            VpcId: 'Test',
          },
        },
      },
    });
  });
});
