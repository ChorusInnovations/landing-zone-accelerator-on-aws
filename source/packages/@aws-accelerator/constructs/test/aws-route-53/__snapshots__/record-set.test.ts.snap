// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecordSet Construct(RecordSet):  Snapshot Test 1`] = `
Object {
  "Resources": Object {
    "TestHostedZone68F306E4": Object {
      "Properties": Object {
        "Name": Object {
          "Fn::Join": Array [
            "",
            Array [
              "s3-global.accesspoint.",
              Object {
                "Ref": "AWS::Region",
              },
              ".amazonaws.com",
            ],
          ],
        },
        "VPCs": Array [
          Object {
            "VPCId": Object {
              "Ref": "TestVpcE77CE678",
            },
            "VPCRegion": Object {
              "Ref": "AWS::Region",
            },
          },
        ],
      },
      "Type": "AWS::Route53::HostedZone",
    },
    "TestRecordSetED81F5C1": Object {
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": Object {
            "Fn::Select": Array [
              1,
              Object {
                "Fn::Split": Array [
                  ":",
                  Object {
                    "Fn::Select": Array [
                      0,
                      Object {
                        "Fn::GetAtt": Array [
                          "TestVpcEndpointF7CADE71",
                          "DnsEntries",
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
          "HostedZoneId": Object {
            "Fn::Select": Array [
              0,
              Object {
                "Fn::Split": Array [
                  ":",
                  Object {
                    "Fn::Select": Array [
                      0,
                      Object {
                        "Fn::GetAtt": Array [
                          "TestVpcEndpointF7CADE71",
                          "DnsEntries",
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        },
        "HostedZoneId": Object {
          "Ref": "TestHostedZone68F306E4",
        },
        "Name": Object {
          "Fn::Join": Array [
            "",
            Array [
              "s3-global.accesspoint.",
              Object {
                "Ref": "AWS::Region",
              },
              ".amazonaws.com",
            ],
          ],
        },
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "TestRouteTableD3D1979F": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Network-Endpoints-Default",
          },
        ],
        "VpcId": Object {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "TestSecurityGroupDA4B5F83": Object {
      "Properties": Object {
        "GroupDescription": "AWS Private Endpoint Zone",
        "GroupName": "TestSecurityGroup",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestSecurityGroup",
          },
        ],
        "VpcId": Object {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "TestSubnet7ABA0E56": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Join": Array [
            "",
            Array [
              Object {
                "Ref": "AWS::Region",
              },
              "a",
            ],
          ],
        },
        "CidrBlock": "10.1.0.0/24",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Network-Endpoints-A",
          },
        ],
        "VpcId": Object {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "TestSubnetRouteTableAssociationFE267B30": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "TestRouteTableD3D1979F",
        },
        "SubnetId": Object {
          "Ref": "TestSubnet7ABA0E56",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "TestVpcE77CE678": Object {
      "Properties": Object {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": false,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Test",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "TestVpcEndpointF7CADE71": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "*",
              "Condition": Object {
                "StringEquals": Object {
                  "aws:PrincipalOrgID": Array [
                    "organizationId",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": "*",
              "Sid": "AccessToTrustedPrincipalsAndResources",
            },
          ],
          "Version": "2012-10-17",
        },
        "PrivateDnsEnabled": false,
        "SecurityGroupIds": Array [
          Object {
            "Ref": "TestSecurityGroupDA4B5F83",
          },
        ],
        "ServiceName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "com.amazonaws.",
              Object {
                "Ref": "AWS::Region",
              },
              ".ec2",
            ],
          ],
        },
        "SubnetIds": Array [
          Object {
            "Ref": "TestSubnet7ABA0E56",
          },
        ],
        "VpcEndpointType": "Interface",
        "VpcId": Object {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "TestVpcInternetGateway01360C82": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Test",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "TestVpcInternetGatewayAttachment60E451D5": Object {
      "Properties": Object {
        "InternetGatewayId": Object {
          "Ref": "TestVpcInternetGateway01360C82",
        },
        "VpcId": Object {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
}
`;
